corehq/apps/userreports/expressions/json_query_specs.py

```diff
@@ -0,0 +1,98 @@
+import itertools
+from dimagi.ext.jsonobject import JsonObject, DictProperty, StringProperty
+from corehq.apps.userreports.exceptions import BadSpecError
+from corehq.apps.userreports.specs import TypeProperty
+from jsonobject.base_properties import DefaultProperty
+from .utils import SUPPORTED_UCR_AGGREGATIONS, aggregate_items
+
+
+class FilterItemsExpressionSpec(JsonObject):
+    type = TypeProperty('filter_items')
+    items_expression = DefaultProperty(required=True)
+    filter_expression = DictProperty(required=True)
+
+    def configure(self, items_expression, filter_expression):
+        self._items_expression = items_expression
+        self._filter_expression = filter_expression
+
+    def __call__(self, doc, context=None):
+        items = self._items_expression(doc, context) or []
+
+        values = []
+        for item in items:
+            if self._filter_expression(item, context):
+                values.append(item)
+
+        return values
+
+
+class MapItemsExpressionSpec(JsonObject):
+    type = TypeProperty('map_items')
+    items_expression = DefaultProperty(required=True)
+    map_expression = DictProperty(required=True)
+
+    def configure(self, items_expression, map_expression):
+        self._items_expression = items_expression
+        self._map_expression = map_expression
+
+    def __call__(self, doc, context=None):
+        items = self._items_expression(doc, context) or []
+
+        return map(
```

Comment:
same comment as above. might want to abstract the notion of getting a list from an expression or failing gracefully to a function.